src/uct/ib/mlx5/gga/gga_mlx5.c

```diff
@@ -268,57 +361,133 @@ static uct_component_t uct_gga_component = {
 };
 
 static UCS_F_ALWAYS_INLINE void
- uct_gga_mlx5_rkey_trace(uct_gga_mlx5_md_t *md,
-                         uct_gga_mlx5_rkey_handle_t *rkey_handle,
-                         const char *prefix)
+uct_gga_mlx5_rkey_trace(uct_gga_mlx5_md_t *md,
+                        uct_gga_mlx5_rkey_handle_t *rkey_handle,
+                        const char *prefix)
 {
-    ucs_trace("md %p: %s resolved rkey %p: rkey_ob %"PRIx64"/%p", md, prefix,
-              rkey_handle, rkey_handle->rkey_ob.rkey,
-              rkey_handle->rkey_ob.handle);
+    ucs_trace("md %p: %s resolved rkey_handle %p: rkey %"PRIx64, md, prefix,
+              rkey_handle, rkey_handle->rkey);
 }
 
-static UCS_F_ALWAYS_INLINE ucs_status_t
-uct_gga_mlx5_rkey_resolve(uct_gga_mlx5_md_t *md,
-                          uct_gga_mlx5_rkey_handle_t *rkey_handle)
+/**
+ * A caller must hold @a uct_gga_mlx5_rkey_handle_t::lock
+ */
+static void
+uct_gga_mlx5_rkey_handle_update(uct_gga_mlx5_rkey_handle_t *rkey_handle,
+                                const uct_gga_mlx5_md_t *md,
+                                const uct_gga_mlx5_rkey_bundle_t *rkey_bundle)
 {
-    uct_md_h uct_md                          = &md->super.super.super;
-    uct_md_mem_attach_params_t attach_params = { 0 };
-    uct_md_mkey_pack_params_t repack_params  = { 0 };
+    rkey_handle->md   = md;
+    rkey_handle->rkey = rkey_bundle->rkey_ob.rkey;
+    ucs_assert(rkey_handle->rkey != UCT_INVALID_RKEY);
+}
+
+static ucs_status_t
+uct_gga_mlx5_rkey_resolve_slow(const uct_gga_mlx5_rkey_hash_key_t *hash_key,
+                               uct_rkey_t *rkey)
+{
+    uct_gga_mlx5_rkey_handle_t *rkey_handle = hash_key->rkey_handle;
+    uct_gga_mlx5_md_t *md                   = hash_key->md;
+    uct_md_h uct_md                         = &md->super.super.super;
+    uct_gga_mlx5_rkey_bundle_t *rkey_bundle;
+    uct_md_mem_attach_params_t attach_params;
+    uct_md_mkey_pack_params_t repack_params;
     uint64_t repack_mkey;
     ucs_status_t status;
 
-    if (ucs_likely(rkey_handle->memh != NULL)) {
-        uct_gga_mlx5_rkey_trace(md, rkey_handle, "reuse");
-        return UCS_OK;
+    rkey_bundle = ucs_malloc(sizeof(*rkey_bundle), "gga_rkey_bundle");
```

Comment:
You can avoid malloc/free if you use uct_gga_mlx5_rkey_bundle_t as hash map value instead of uct_gga_mlx5_rkey_bundle_t*.
khash implementation already has malloc/free for hash map values. So after kh_put, you can initialize the value.